File: test/test_did_key.cpp
#include <array>

#include <doctest/doctest.h>

#include <authbox.hpp>
#include <pki/key_utils.hpp>

TEST_SUITE("did/key") {

    TEST_CASE("encode and parse did:key ed25519") {
        auto keypair = authbox::pik::generate_ed25519_keypair();
        REQUIRE(keypair.public_key.size() == 32);

        std::array<uint8_t, 32> pub{};
        std::copy_n(keypair.public_key.begin(), 32, pub.begin());

        auto did = authbox::did::encode_ed25519_did_key(pub);
        REQUIRE(did.is_ok());
        CHECK(std::string_view(did.value()).starts_with("did:key:z"));

        auto parsed = authbox::did::parse_did_key(did.value());
        REQUIRE(parsed.is_ok());
        CHECK(parsed.value().type == authbox::did::DidKeyType::ED25519);
        CHECK(parsed.value().public_key == keypair.public_key);
    }

    TEST_CASE("resolve did:key document") {
        auto keypair = authbox::pik::generate_ed25519_keypair();
        std::array<uint8_t, 32> pub{};
        std::copy_n(keypair.public_key.begin(), 32, pub.begin());

        auto did = authbox::did::encode_ed25519_did_key(pub);
        REQUIRE(did.is_ok());

        auto json = authbox::did::resolve_did_key_document_json(did.value());
        REQUIRE(json.is_ok());

        auto parsed_doc = authbox::did::parse_document(json.value());
        REQUIRE(parsed_doc.is_ok());
        CHECK(std::string_view(parsed_doc.value().id.data(), parsed_doc.value().id.size()) == did.value());
        CHECK(parsed_doc.value().verification_methods.size() == 1);
        CHECK(parsed_doc.value().authentication.size() == 1);
    }

    TEST_CASE("resolver supports did:key without network fetcher") {
        auto keypair = authbox::pik::generate_ed25519_keypair();
        std::array<uint8_t, 32> pub{};
        std::copy_n(keypair.public_key.begin(), 32, pub.begin());
        auto did = authbox::did::encode_ed25519_did_key(pub);
        REQUIRE(did.is_ok());

        authbox::did::Resolver resolver;
        auto resolved = resolver.resolve(did.value());
        REQUIRE(resolved.is_ok());
        CHECK(std::string_view(resolved.value().source_url.data(), resolved.value().source_url.size()) ==
              "did:key:inline");
        CHECK(std::string_view(resolved.value().document.id.data(), resolved.value().document.id.size()) ==
              did.value());
    }
}

File: src/did/tests/test_did_key.rs
use super::super::*;

#[test]
fn did_key_encodes_and_parses_ed25519_keys() {
    let keypair = crate::pki::generate_ed25519_keypair().unwrap();
    assert_eq!(keypair.public_key.len(), 32);

    let public_key: [u8; 32] = keypair.public_key.as_slice().try_into().unwrap();
    let did = encode_ed25519_did_key(public_key).unwrap();

    assert!(did.starts_with("did:key:z"));
    let parsed = parse_did_key(&did).unwrap();
    assert_eq!(parsed.type_, DidKeyType::Ed25519);
    assert_eq!(parsed.public_key, keypair.public_key);
}

#[test]
fn did_key_cpp_named_type_constants_match_rust_variants() {
    assert_eq!(DidKeyType::ED25519, DidKeyType::Ed25519);
    assert_eq!(DidKeyType::X25519, DidKeyType::X25519);
}

#[test]
fn did_key_info_default_matches_cpp_value_initialized_aggregate() {
    let info = DidKeyInfo::default();

    assert_eq!(info.type_, DidKeyType::ED25519);
    assert!(info.public_key.is_empty());
    assert!(info.prefixed_key.is_empty());
    assert_eq!(info.fingerprint, "");
    assert_eq!(info.did_uri, "");
}

#[test]
fn did_key_resolves_document_json() {
    let keypair = crate::pki::generate_ed25519_keypair().unwrap();
    let public_key: [u8; 32] = keypair.public_key.as_slice().try_into().unwrap();
    let did = encode_ed25519_did_key(public_key).unwrap();

    let json = resolve_did_key_document_json(&did).unwrap();
    let document = parse_document(&json).unwrap();

    assert_eq!(document.id, did);
    assert_eq!(document.verification_methods.len(), 1);
    assert_eq!(document.authentication.len(), 1);
}

#[test]
fn did_key_resolver_supports_inline_resolution_without_network_fetcher() {
    let keypair = crate::pki::generate_ed25519_keypair().unwrap();
    let public_key: [u8; 32] = keypair.public_key.as_slice().try_into().unwrap();
    let did = encode_ed25519_did_key(public_key).unwrap();

    let resolver = Resolver::new();
    let resolved = resolver.resolve(&did).unwrap();

    assert_eq!(resolved.source_url, "did:key:inline");
    assert_eq!(resolved.document.id, did);
}
